# Data Loading: 


### Importing Libraries:

In [6]:
# Importing the pandas library
import pandas as pd

### Reading dataset Error:

In [7]:
# Creating a data frame 
# sample_superstore = pd.read_csv("Original_dataset/Sample-Superstore.csv")

When we try to read the csv file we get this error : 'UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 2944: invalid star byte' it's mean that the default encoding 'utf-8'is not the appropriate one for reading the csv file. 

### Encoding function for reading dataset:

Create a function they try witch encoding work to read our data :

In [8]:
# Creating a function that give us witch encoding can we use to read the csv file with pandas
def read_csv_with_encoding(filename):
    """
    Function to read a CSV file with different encoding options.
    
    Parameters:
        filename (str): The path to the CSV file.
        
    Returns:
        DataFrame or None: Returns DataFrame if reading is successful, else returns None.
    """
    encoding_list=['utf-8', 'latin1', 'iso-8859-1', 'windows-1252', 'latin2', 'utf-16le', 'utf-16be', 'ascii', 'cp1252']
    
    for encoding in encoding_list:
        try:
            df = pd.read_csv(filename, encoding=encoding)
            print(f"CSV file successfully read with encoding: {encoding}")
        except Exception as e:
            print(f"Failed to read CSV with encoding '{encoding}': {e}")
    
    return None

In [9]:
# Creat a variable with the path or the csv file we want to import 
csv_file = "Original_dataset/Sample-Superstore.csv"

# Use the function to get witch encoding we can use
read_csv_with_encoding(csv_file)

Failed to read CSV with encoding 'utf-8': 'utf-8' codec can't decode byte 0xf6 in position 9: invalid start byte
CSV file successfully read with encoding: latin1
CSV file successfully read with encoding: iso-8859-1
CSV file successfully read with encoding: windows-1252
CSV file successfully read with encoding: latin2
CSV file successfully read with encoding: utf-16le
CSV file successfully read with encoding: utf-16be
Failed to read CSV with encoding 'ascii': 'ascii' codec can't decode byte 0xa0 in position 2944: ordinal not in range(128)
CSV file successfully read with encoding: cp1252


### Reading dataset:

In [10]:
# Creating a data frame 
sample_superstore = pd.read_csv("Original_dataset/Sample-Superstore.csv", encoding='latin1')

# Pre-Processing of data and Initial Exploration:

### Dimensions of the data:

In [11]:
#Display the total numbers of rows and columns
display(sample_superstore.shape)

(9994, 21)

### Peek at the data:

In [12]:
# Display the 5 first rows of the data
display(sample_superstore.head())

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
1       2  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
3       4  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [13]:
# Display the 5 last rows of the data 
sample_superstore.tail()

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
9989    9990  CA-2014-110422  1/21/2014  1/23/2014    Second Class   
9990    9991  CA-2017-121258  2/26/2017   3/3/2017  Standard Class   
9991    9992  CA-2017-121258  2/26/2017   3/3/2017  Standard Class   
9992    9993  CA-2017-121258  2/26/2017   3/3/2017  Standard Class   
9993    9994  CA-2017-119914   5/4/2017   5/9/2017    Second Class   

     Customer ID     Customer Name   Segment        Country         City  ...  \
9989    TB-21400  Tom Boeckenhauer  Consumer  United States        Miami  ...   
9990    DB-13060       Dave Brooks  Consumer  United States   Costa Mesa  ...   
9991    DB-13060       Dave Brooks  Consumer  United States   Costa Mesa  ...   
9992    DB-13060       Dave Brooks  Consumer  United States   Costa Mesa  ...   
9993    CC-12220      Chris Cortes  Consumer  United States  Westminster  ...   

     Postal Code  Region       Product ID         Category Sub-Category  \
9989       33180   South  FUR-FU-10001889        Furniture  Furnishings   
9990       92627    West  FUR-FU-10000747        Furniture  Furnishings   
9991       92627    West  TEC-PH-10003645       Technology       Phones   
9992       92627    West  OFF-PA-10004041  Office Supplies        Paper   
9993       92683    West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name    Sales  Quantity  \
9989                             Ultra Door Pull Handle   25.248         3   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.960         2   
9991                              Aastra 57i VoIP phone  258.576         2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.600         4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.160         2   

      Discount   Profit  
9989       0.2   4.1028  
9990       0.0  15.6332  
9991       0.2  19.3932  
9992       0.0  13.3200  
9993       0.0  72.9480  

[5 rows x 21 columns]

### Columns in the data:

In [14]:
# Display the columns names 
sample_superstore.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

We need to change the names of the columns

### Data informations:

In [15]:
# Display information about the data frame 
sample_superstore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

We gonna change the datatypes of the columns

### Duplicated rows:

In [16]:
# Display the duplicated rows 
sample_superstore.duplicated().sum()

0

### None Values in the data:

In [17]:
# Display the none values 
sample_superstore.isna().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

# Cleaning Data: 

Making an copy :


In [18]:
# Creating a copy of the original data frame before any modifications
clean_df = sample_superstore.copy()

### Rename columns:

In [19]:
# Let's change the names of all columns in the data frame 

# Creating a list with all updates columns names
new_columns_name = ['order_number', 'order_id', 'order_date', 'ship_date', 'ship_mode',
        'customer_id', 'customer_name', 'segment', 'country', 'city', 'state',
        'postal_code', 'region', 'product_id', 'category', 'sub_category',
        'product_name', 'sales', 'quantity', 'discount', 'profit']

# Assign the new columns names to the data frame 
clean_df.columns = new_columns_name

# Display the info of the dataframe 
display(clean_df.columns)

Index(['order_number', 'order_id', 'order_date', 'ship_date', 'ship_mode',
       'customer_id', 'customer_name', 'segment', 'country', 'city', 'state',
       'postal_code', 'region', 'product_id', 'category', 'sub_category',
       'product_name', 'sales', 'quantity', 'discount', 'profit'],
      dtype='object')

### Converting columns data types:

In [20]:
# Let's change the columns datatype of order_date ans shipe_date

# Coverting the two columns to datetime datatype
clean_df['order_date'] = pd.to_datetime(clean_df['order_date'])
clean_df['ship_date'] = pd.to_datetime(clean_df['ship_date'])


In [21]:
# Display the colummns datatypes
clean_df.dtypes

order_number              int64
order_id                 object
order_date       datetime64[ns]
ship_date        datetime64[ns]
ship_mode                object
customer_id              object
customer_name            object
segment                  object
country                  object
city                     object
state                    object
postal_code               int64
region                   object
product_id               object
category                 object
sub_category             object
product_name             object
sales                   float64
quantity                  int64
discount                float64
profit                  float64
dtype: object

## Checking if they have some incoherences in the data frame:

### Date:

In [22]:
# Min and Max from the 'order_date'
order = clean_df['order_date'].min(), clean_df['order_date'].max()
# Min and Max from the 'ship_date'
ship = clean_df['ship_date'].min(), clean_df['ship_date'].max()

# Print the results of the 2 variables
print(f"Min and Max order date: {order}")
print(f"Min and Max ship date: {ship}")

Min and Max order date: (Timestamp('2014-01-03 00:00:00'), Timestamp('2017-12-30 00:00:00'))
Min and Max ship date: (Timestamp('2014-01-07 00:00:00'), Timestamp('2018-01-05 00:00:00'))


In [23]:
# Check if 'order_date' values are strict less than 'ship_order' values

# Creating a variable with the condition 
date_comparison = clean_df['order_date'] > clean_df['ship_date']

# Display the rows where the condition is True 
clean_df[date_comparison]

Empty DataFrame
Columns: [order_number, order_id, order_date, ship_date, ship_mode, customer_id, customer_name, segment, country, city, state, postal_code, region, product_id, category, sub_category, product_name, sales, quantity, discount, profit]
Index: []

[0 rows x 21 columns]

We didn't have any problem with the date


### Unique customer id by customer name:

In [24]:
# Group by 'customer_id' and count the number of unique 'name' values in each group
customer_id_counts = clean_df.groupby('customer_id')['customer_name'].nunique()

# Check if any customer has more than one unique 'name' value (customer ID)
customers_with_multiple_ids = customer_id_counts[customer_id_counts > 1]

# If customers_with_multiple_ids is empty, it means each customer has only one ID
if customers_with_multiple_ids.empty:
    print("Each customer has only one customer ID.")
else:
    print("The following customers have multiple customer IDs:")
    print(customers_with_multiple_ids)

Each customer has only one customer ID.


### Unique postal code by city:

In [25]:
# Group by 'poste_code' and count the number of unique 'city' values in each group
poste_code_counts = clean_df.groupby('postal_code')['city'].nunique()

# Check if any poste code has more than one unique 'city' value
poste_code_with_multiple_city = poste_code_counts[poste_code_counts > 1]

# If poste_code_with_multiple_city is empty, it means each poste code has only one city
if poste_code_with_multiple_city.empty:
    print("Each city has only one poste code.")
else:
    print("The following poste code have multiple city:")
    print(poste_code_with_multiple_city)

The following poste code have multiple city:
postal_code
92024    2
Name: city, dtype: int64


We have 2 city with the same postal code, we gonna fix that!

### Unique product id by product name:

In [26]:
# Group by 'product_id' and count the number of unique 'name' values in each group
product_id_counts = clean_df.groupby('product_id')['product_name'].nunique()

# Check if any product has more than one unique 'name' value (product ID)
products_with_multiple_ids = product_id_counts[product_id_counts > 1]

# If products_with_multiple_ids is empty, it means each product has only one ID
if products_with_multiple_ids.empty:
    print("Each product has only one product ID.")
else:
    print("The following product have multiple product IDs:")
    print(products_with_multiple_ids)
    
print(products_with_multiple_ids.value_counts())

The following product have multiple product IDs:
product_id
FUR-BO-10002213    2
FUR-CH-10001146    2
FUR-FU-10001473    2
FUR-FU-10004017    2
FUR-FU-10004091    2
FUR-FU-10004270    2
FUR-FU-10004848    2
FUR-FU-10004864    2
OFF-AP-10000576    2
OFF-AR-10001149    2
OFF-BI-10002026    2
OFF-BI-10004632    2
OFF-BI-10004654    2
OFF-PA-10000357    2
OFF-PA-10000477    2
OFF-PA-10000659    2
OFF-PA-10001166    2
OFF-PA-10001970    2
OFF-PA-10002195    2
OFF-PA-10002377    2
OFF-PA-10003022    2
OFF-ST-10001228    2
OFF-ST-10004950    2
TEC-AC-10002049    2
TEC-AC-10002550    2
TEC-AC-10003832    2
TEC-MA-10001148    2
TEC-PH-10001530    2
TEC-PH-10001795    2
TEC-PH-10002200    2
TEC-PH-10002310    2
TEC-PH-10004531    2
Name: product_name, dtype: int64
product_name
2    32
Name: count, dtype: int64


How we can see, we have 32 id associated with the same product name, we gonna try to fix that.

# Data Cleaning:

## Fixe the incoherence of the data frame:

Create a copy of the clean dataframe before removing any values:

In [27]:
# Create a copy of clean_df
fixed_df = clean_df.copy()

### Fixe the incoherence Postal Code:

We gonna first check witch City have the postal_code = 92024

In [28]:
# Filter the dataframe for rows with the postal code 92024
postal_code = fixed_df[fixed_df['postal_code'] == 92024]
# Uniaue cities associated with this postal code 
unique_city = postal_code['city'].unique()

# Print the cities
print(unique_city)

['San Diego' 'Encinitas']


After checking on internet, the city associated to the postal code 92024 is Encinitas. Now we need to found in the dataframe witch postal code is associated to the city San Diego.

In [29]:
# Filter the dataframe for rows with the city name San Diego
city_name = fixed_df[fixed_df['city'] == 'San Diego']
# Uniaue cities associated with this postal code 
unique_postal_code = city_name['postal_code'].unique()

# Print the posrtal code 
print(unique_postal_code)

[92037 92024 92105]


How we can see with have 2 postal code (92037, 92105) associated to the city San Diego. So we gonna choose one of this postal code and replace the wrong code for San Diego by the correct one. 

In [30]:
# Update the postal code 92024 to 92037 for all values where the city is San Diego
fixed_df.loc[(fixed_df['postal_code'] == 92024) & (fixed_df['city'] == 'San Diego'), 'city'] = 92037

Now we gonna check if we have only this 2 Postal code associated to the city San Diego

In [31]:
# Filter the dataframe for rows with the city name San Diego
city_name = fixed_df[fixed_df['city'] == 'San Diego']
# Uniaue cities associated with this postal code 
unique_postal_code = city_name['postal_code'].unique()

# Print the posrtal code 
print(unique_postal_code)

[92037 92105]


In [32]:
# Numbers of unique Postal Code 
fixed_df['postal_code'].nunique()

631

## Fixe the incoherence of Product:

### Unique product IDs having multiple name:

In [33]:
# Group by 'product_id' and count the number of unique 'name' values in each group
product_id_counts = fixed_df.groupby('product_id')['product_name'].nunique()

# Check if any product has more than one unique 'name' value (product ID)
products_with_multiple_ids = product_id_counts[product_id_counts > 1]

# If products_with_multiple_ids is empty, it means each product has only one ID
if products_with_multiple_ids.empty:
    print("Each product has only one product ID.")
else:
    print("The following product have multiple product IDs:")
    print(products_with_multiple_ids)

The following product have multiple product IDs:
product_id
FUR-BO-10002213    2
FUR-CH-10001146    2
FUR-FU-10001473    2
FUR-FU-10004017    2
FUR-FU-10004091    2
FUR-FU-10004270    2
FUR-FU-10004848    2
FUR-FU-10004864    2
OFF-AP-10000576    2
OFF-AR-10001149    2
OFF-BI-10002026    2
OFF-BI-10004632    2
OFF-BI-10004654    2
OFF-PA-10000357    2
OFF-PA-10000477    2
OFF-PA-10000659    2
OFF-PA-10001166    2
OFF-PA-10001970    2
OFF-PA-10002195    2
OFF-PA-10002377    2
OFF-PA-10003022    2
OFF-ST-10001228    2
OFF-ST-10004950    2
TEC-AC-10002049    2
TEC-AC-10002550    2
TEC-AC-10003832    2
TEC-MA-10001148    2
TEC-PH-10001530    2
TEC-PH-10001795    2
TEC-PH-10002200    2
TEC-PH-10002310    2
TEC-PH-10004531    2
Name: product_name, dtype: int64


### Creating a dictionary of Product ID choosing the first Product Name:

In [34]:
# Get the product IDs with multiple names
products_with_multiple_ids = product_id_counts[product_id_counts > 1].index

# Create dictionary mapping each product ID to its first product name
first_product_names = fixed_df[fixed_df['product_id'].isin(products_with_multiple_ids)].groupby('product_id')['product_name'].first().to_dict()

# Print the dictionary
display(first_product_names)

# Total numbers of Products ID and names
display(len(first_product_names))

{'FUR-BO-10002213': 'DMI Eclipse Executive Suite Bookcases',
 'FUR-CH-10001146': "Global Value Mid-Back Manager's Chair, Gray",
 'FUR-FU-10001473': 'DAX Wood Document Frame',
 'FUR-FU-10004017': 'Tenex Contemporary Contur Chairmats for Low and Medium Pile Carpet, Computer, 39" x 49"',
 'FUR-FU-10004091': 'Howard Miller 13" Diameter Goldtone Round Wall Clock',
 'FUR-FU-10004270': 'Eldon Image Series Desk Accessories, Burgundy',
 'FUR-FU-10004848': 'Howard Miller 13-3/4" Diameter Brushed Chrome Round Wall Clock',
 'FUR-FU-10004864': 'Howard Miller 14-1/2" Diameter Chrome Round Wall Clock',
 'OFF-AP-10000576': 'Belkin 7 Outlet SurgeMaster II',
 'OFF-AR-10001149': 'Sanford Colorific Colored Pencils, 12/Box',
 'OFF-BI-10002026': 'Avery Arch Ring Binders',
 'OFF-BI-10004632': 'Ibico Hi-Tech Manual Binding System',
 'OFF-BI-10004654': 'Avery Binding System Hidden Tab Executive Style Index Sets',
 'OFF-PA-10000357': 'White Dual Perf Computer Printout Paper, 2700 Sheets, 1 Part, Heavyweight, 20

32

### Update the column Product Name with the values in the dictionary:

In [35]:
# Update the 'product_name' column to keep only the first product name for each product ID
for id_product, name_product in first_product_names.items():
    fixed_df.loc[fixed_df['product_id'] == id_product, 'product_name'] = name_product

### Verification:

We gonna verify if the Product ID 'TEC-PH-10002200' is associated to the Product Name 'Samsung Galaxy Note 2'

In [36]:
fixed_df[fixed_df['product_id'] == 'TEC-PH-10002200']

order_number        order_id order_date  ship_date       ship_mode  \
1219          1220  US-2017-118087 2017-09-09 2017-09-13  Standard Class   
1805          1806  US-2016-116729 2016-12-25 2016-12-28     First Class   
3126          3127  CA-2017-149895 2017-09-08 2017-09-12  Standard Class   
3218          3219  CA-2017-104640 2017-11-10 2017-11-16  Standard Class   
4714          4715  CA-2017-121643 2017-05-18 2017-05-20     First Class   
4990          4991  CA-2017-107321 2017-08-31 2017-09-04  Standard Class   
7952          7953  CA-2015-145814 2015-11-19 2015-11-24  Standard Class   
9193          9194  CA-2015-141810 2015-11-02 2015-11-07  Standard Class   

     customer_id     customer_name      segment        country           city  \
1219    SP-20620  Stefania Perrino    Corporate  United States   Philadelphia   
1805    GK-14620       Grace Kelly    Corporate  United States    Los Angeles   
3126    EB-14110    Eugene Barchas     Consumer  United States   Philadelphia   
3218    FH-14275      Frank Hawley    Corporate  United States  New York City   
4714    AB-10105     Adrian Barton     Consumer  United States       Portland   
4990    AW-10930   Arthur Wiediger  Home Office  United States  San Francisco   
7952    KD-16345  Katherine Ducich     Consumer  United States  New York City   
9193    BB-10990   Barry Blumstein    Corporate  United States    San Antonio   

      ... postal_code   region       product_id    category sub_category  \
1219  ...       19134     East  TEC-PH-10002200  Technology       Phones   
1805  ...       90049     West  TEC-PH-10002200  Technology       Phones   
3126  ...       19134     East  TEC-PH-10002200  Technology       Phones   
3218  ...       10024     East  TEC-PH-10002200  Technology       Phones   
4714  ...       97206     West  TEC-PH-10002200  Technology       Phones   
4990  ...       94110     West  TEC-PH-10002200  Technology       Phones   
7952  ...       10035     East  TEC-PH-10002200  Technology       Phones   
9193  ...       78207  Central  TEC-PH-10002200  Technology       Phones   

               product_name     sales  quantity  discount    profit  
1219  Samsung Galaxy Note 2  1931.958         7       0.4 -386.3916  
1805  Samsung Galaxy Note 2  2575.944         7       0.2  257.5944  
3126  Samsung Galaxy Note 2   258.528         2       0.4  -47.3968  
3218  Samsung Galaxy Note 2   430.880         2       0.0  124.9552  
4714  Samsung Galaxy Note 2   344.704         2       0.2   38.7792  
4990  Samsung Galaxy Note 2   689.408         4       0.2   77.5584  
7952  Samsung Galaxy Note 2   861.760         4       0.0  249.9104  
9193  Samsung Galaxy Note 2   344.704         2       0.2   38.7792  

[8 rows x 21 columns]

Let's check again if we didn't have any Product ID with multiple Product Names:

In [37]:
# Group by 'product_id' and count the number of unique 'name' values in each group
product_id_counts = fixed_df.groupby('product_id')['product_name'].nunique()

# Check if any product has more than one unique 'name' value (product ID)
products_with_multiple_ids = product_id_counts[product_id_counts > 1]

# If products_with_multiple_ids is empty, it means each product has only one ID
if products_with_multiple_ids.empty:
    print("Each product has only one product ID.")
else:
    print("The following product have multiple product IDs:")
    print(products_with_multiple_ids)


Each product has only one product ID.


In [38]:
# Total numbers of unique product_id
fixed_df['product_id'].nunique()

1862

### Unique Product Name with mutiple Product ID:

In [39]:
# Group by 'product_name' and count the number of unique 'id' values in each group
product_ids_counts = fixed_df.groupby('product_name')['product_id'].nunique()

# Check if any product has more than one unique 'id' value (product ID)
products_with_multiple_names = product_ids_counts[product_ids_counts > 1]

# If products_with_multiple_names is empty, it means each product has only one ID
if products_with_multiple_names.empty:
    print("Each product has only one product Name.")
else:
    print("The following product have multiple product Names:")
    print(products_with_multiple_names)

The following product have multiple product Names:
product_name
#10- 4 1/8" x 9 1/2" Recycled Envelopes           2
Avery Non-Stick Binders                           2
Easy-staple paper                                 8
Eldon Wave Desk Accessories                       2
KI Adjustable-Height Table                        2
Okidata C610n Printer                             2
Peel & Seel Recycled Catalog Envelopes, Brown     2
Prang Drawing Pencil Set                          2
Staple envelope                                   9
Staple holder                                     3
Staple magnet                                     2
Staple remover                                    3
Staple-based wall hangings                        2
Staples                                          10
Staples in misc. colors                           7
Storex Dura Pro Binders                           2
Name: product_id, dtype: int64


### Update the column Product ID with the values in the dictionary:

In [40]:
# Get the product Names with multiple products IDs
products_with_multiple_names = product_ids_counts[product_ids_counts > 1].index

# Create dictionary mapping each product Name to its first product ID
first_product_ids = fixed_df[fixed_df['product_name'].isin(products_with_multiple_names)].groupby('product_name')['product_id'].first().to_dict()

# Print the dictionary 
display(first_product_ids)

{'#10- 4 1/8" x 9 1/2" Recycled Envelopes': 'OFF-EN-10000461',
 'Avery Non-Stick Binders': 'OFF-BI-10004140',
 'Easy-staple paper': 'OFF-PA-10000249',
 'Eldon Wave Desk Accessories': 'FUR-FU-10000023',
 'KI Adjustable-Height Table': 'FUR-TA-10004086',
 'Okidata C610n Printer': 'TEC-MA-10003230',
 'Peel & Seel Recycled Catalog Envelopes, Brown': 'OFF-EN-10003448',
 'Prang Drawing Pencil Set': 'OFF-AR-10003251',
 'Staple envelope': 'OFF-EN-10001990',
 'Staple holder': 'OFF-AP-10004249',
 'Staple magnet': 'OFF-ST-10003221',
 'Staple remover': 'OFF-SU-10001225',
 'Staple-based wall hangings': 'FUR-FU-10004351',
 'Staples': 'OFF-FA-10002780',
 'Staples in misc. colors': 'OFF-AR-10001166',
 'Storex Dura Pro Binders': 'OFF-BI-10001922'}

### Update the column Product ID with the values in the dictionary:

In [41]:
# Update the 'product_id' column to keep only the first product ID for each product Name 
for names_product, ids_product in first_product_ids.items():
    fixed_df.loc[fixed_df['product_name'] == names_product, 'product_id'] = ids_product

#### Verification:

We gonna verify if the Product Name 'Staples' is associated to the Product ID 'OFF-FA-10002780'

In [42]:
fixed_df[fixed_df['product_id'] == 'OFF-FA-10002780']

order_number        order_id order_date  ship_date       ship_mode  \
132            133  US-2017-164147 2017-02-02 2017-02-05     First Class   
342            343  CA-2014-122336 2014-04-13 2014-04-17    Second Class   
525            526  CA-2015-158792 2015-12-26 2016-01-02  Standard Class   
646            647  CA-2016-103947 2016-04-01 2016-04-08  Standard Class   
1108          1109  US-2016-110156 2016-11-19 2016-11-24  Standard Class   
1147          1148  CA-2015-112452 2015-04-04 2015-04-04        Same Day   
1485          1486  CA-2014-119032 2014-11-27 2014-12-03  Standard Class   
1642          1643  US-2014-134712 2014-11-29 2014-12-04  Standard Class   
1703          1704  CA-2014-139857 2014-02-02 2014-02-06  Standard Class   
1886          1887  CA-2014-140165 2014-05-13 2014-05-15     First Class   
2059          2060  CA-2014-106439 2014-10-31 2014-11-04  Standard Class   
2144          2145  CA-2016-155481 2016-09-01 2016-09-05  Standard Class   
2387          2388  CA-2015-145065 2015-12-12 2015-12-15     First Class   
2426          2427  US-2017-147669 2017-12-25 2017-12-30  Standard Class   
2701          2702  CA-2014-145317 2014-03-18 2014-03-23  Standard Class   
2923          2924  CA-2014-156993 2014-06-28 2014-07-04  Standard Class   
3095          3096  CA-2015-114468 2015-08-23 2015-08-23        Same Day   
3223          3224  CA-2014-108189 2014-10-02 2014-10-05     First Class   
3274          3275  CA-2017-116358 2017-11-02 2017-11-06  Standard Class   
3345          3346  CA-2016-145982 2016-08-27 2016-09-01    Second Class   
3386          3387  CA-2017-148404 2017-10-07 2017-10-11  Standard Class   
3977          3978  CA-2017-101308 2017-10-02 2017-10-07  Standard Class   
4065          4066  CA-2014-102673 2014-11-01 2014-11-05  Standard Class   
4365          4366  CA-2017-111332 2017-05-20 2017-05-22    Second Class   
4416          4417  CA-2014-153983 2014-11-29 2014-12-06  Standard Class   
4502          4503  CA-2014-116757 2014-06-30 2014-07-04  Standard Class   
4522          4523  CA-2017-105445 2017-02-20 2017-02-25    Second Class   
4555          4556  CA-2015-130218 2015-03-23 2015-03-26     First Class   
4934          4935  CA-2015-106978 2015-09-28 2015-10-04  Standard Class   
4958          4959  CA-2015-127607 2015-03-20 2015-03-26  Standard Class   
5249          5250  CA-2015-106215 2015-04-16 2015-04-21  Standard Class   
5382          5383  CA-2016-149195 2016-09-05 2016-09-07    Second Class   
5678          5679  CA-2016-143924 2016-07-28 2016-08-03  Standard Class   
6130          6131  CA-2017-162033 2017-03-26 2017-04-02  Standard Class   
6206          6207  CA-2016-133697 2016-10-20 2016-10-24    Second Class   
6212          6213  CA-2017-149944 2017-11-12 2017-11-16    Second Class   
6723          6724  US-2017-166394 2017-09-04 2017-09-06    Second Class   
6842          6843  CA-2016-123946 2016-09-12 2016-09-17  Standard Class   
7374          7375  US-2017-123862 2017-01-07 2017-01-09    Second Class   
7776          7777  CA-2014-166961 2014-12-27 2014-12-31  Standard Class   
7899          7900  CA-2017-128363 2017-08-13 2017-08-18  Standard Class   
7956          7957  CA-2017-131807 2017-10-05 2017-10-10  Standard Class   
7983          7984  CA-2017-152499 2017-01-22 2017-01-25    Second Class   
9062          9063  CA-2016-101623 2016-02-28 2016-03-06  Standard Class   
9555          9556  CA-2015-111612 2015-11-28 2015-12-02  Standard Class   
9611          9612  CA-2017-144365 2017-10-24 2017-10-30  Standard Class   

     customer_id     customer_name      segment        country  \
132     DW-13585    Dorothy Wardle    Corporate  United States   
342     JD-15895  Jonathan Doherty    Corporate  United States   
525     BD-11605      Brian Dahlen     Consumer  United States   
646     BB-10990   Barry Blumstein    Corporate  United States   
1108    EH-13945     Eric Hoffmann     Consumer  United States   
1147    NC-18340       Nat Carroll   

Let's verify if we didn't have any multiple product ID for one product Name:

In [43]:
# Group by 'product_name' and count the number of unique 'id' values in each group
product_ids_counts = fixed_df.groupby('product_name')['product_id'].nunique()

# Check if any product has more than one unique 'id' value (product ID)
products_with_multiple_names = product_ids_counts[product_ids_counts > 1]

# If products_with_multiple_names is empty, it means each product has only one ID
if products_with_multiple_names.empty:
    print("Each product has only one product Name.")
else:
    print("The following product have multiple product Names:")
    print(products_with_multiple_names)

Each product has only one product Name.


Let's check if the total of unique product ID it's egal to the total of unique product Name:

In [44]:
fixed_df['product_id'].nunique() == fixed_df['product_name'].nunique()

True

# Order the Data frame by the more recent Order Date: 

Create a copy of the fixed dataframe:

In [45]:
df = fixed_df.copy()

## Drop the Order Number column and Order by order_date:

In [46]:
# Order the Dataframe by 'order_date' in descending order
df = df.drop(columns=['order_number']).sort_values(by='order_date', ascending=False)
# Reset the index
df = df.reset_index(drop=True)

## Add the column 'order_number':

In [47]:
# Add a new column 'order_number' with values counting the rows in descending order
df['order_number'] = range(len(df), 0, -1)

# Reorder the columns to have 'order_number' as the first column
df = df[['order_number'] + [col for col in df.columns if col != 'order_number']]

In [48]:
# Print the update Dataframe
df

order_number        order_id order_date  ship_date       ship_mode  \
0             9994  CA-2017-143259 2017-12-30 2018-01-03  Standard Class   
1             9993  CA-2017-143259 2017-12-30 2018-01-03  Standard Class   
2             9992  CA-2017-115427 2017-12-30 2018-01-03  Standard Class   
3             9991  CA-2017-115427 2017-12-30 2018-01-03  Standard Class   
4             9990  CA-2017-126221 2017-12-30 2018-01-05  Standard Class   
...            ...             ...        ...        ...             ...   
9989             5  CA-2014-141817 2014-01-05 2014-01-12  Standard Class   
9990             4  CA-2014-112326 2014-01-04 2014-01-08  Standard Class   
9991             3  CA-2014-112326 2014-01-04 2014-01-08  Standard Class   
9992             2  CA-2014-112326 2014-01-04 2014-01-08  Standard Class   
9993             1  CA-2014-103800 2014-01-03 2014-01-07  Standard Class   

     customer_id      customer_name      segment        country  \
0       PO-18865  Patrick O'Donnell     Consumer  United States   
1       PO-18865  Patrick O'Donnell     Consumer  United States   
2       EB-13975         Erica Bern    Corporate  United States   
3       EB-13975         Erica Bern    Corporate  United States   
4       CC-12430        Chuck Clark  Home Office  United States   
...          ...                ...          ...            ...   
9989    MB-18085         Mick Brown     Consumer  United States   
9990    PO-19195      Phillina Ober  Home Office  United States   
9991    PO-19195      Phillina Ober  Home Office  United States   
9992    PO-19195      Phillina Ober  Home Office  United States   
9993    DP-13000      Darren Powers     Consumer  United States   

               city  ... postal_code   region       product_id  \
0     New York City  ...       10009     East  FUR-BO-10003441   
1     New York City  ...       10009     East  TEC-PH-10004774   
2         Fairfield  ...       94533     West  OFF-BI-10002103   
3         Fairfield  ...       94533     West  OFF-BI-10004632   
4          Columbus  ...       47201  Central  OFF-AP-10002457   
...             ...  ...         ...      ...              ...   
9989   Philadelphia  ...       19143     East  OFF-AR-10003478   
9990     Naperville  ...       60540  Central  OFF-BI-10004094   
9991     Naperville  ...       60540  Central  OFF-LA-10003223   
9992     Naperville  ...       60540  Central  OFF-ST-10002743   
9993        Houston  ...       77095  Central  OFF-PA-10000174   

             category sub_category  \
0           Furniture    Bookcases   
1          Technology       Phones   
2     Office Supplies      Binders   
3     Office Supplies      Binders   
4     Office Supplies   Appliances   
...               ...          ...   
9989  Office Supplies          Art   
9990  Office Supplies      Binders   
9991  Office Supplies       Labels   
9992  Office Supplies      Storage   
9993  Office Supplies        Paper   

                                           product_name    sales  quantity  \
0     Bush Westfield Collection Bookcases, Fully Ass...  323.136         4   
1                             Gear Head AU3700S Headset   90.930         7   
2       Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl   13.904         2   
3                   Ibico Hi-Tech Manual Binding System   20.720         2   
4     Eureka The Boss Plus 12-Amp Hard Box Upright V...  209.300         2   
...                                                 ...      ...       ...   
9989  Avery Hi-Liter EverBold Pen Style Fluorescent ...   19.536         3   
9990         GBC Standard Plastic Binding Systems Combs    3.540         2   
9991                                          Avery 508   11.784         3   
9992                      SAFCO Boltless Steel Shelving  272.736         3   
9993  Message Book, Wirebound, Four 5 1/2" X 4" Form...   16.448         2   

      discount   profit  
0          0.2  12.1176  
1          0.0   2.7279  
2          0.2   4.5188

# Exporting the clean dataframe in to CSV file:

In [49]:
# Export Dataframe to a CSV file 
#df.to_csv('SuperStore.csv', index=False)

# Print a message to confirm the exportation
print("Datafreme exported successfully!")

Datafreme exported successfully!
